servicetest/cgroups/test_cgroups.py

```diff
@@ -0,0 +1,178 @@
+                        ])
+
+manifest = StandardManifest(TESTOUTPUT_DIR,
+                            "cgroup-test-manifest.json",
+                            [test_cap_0, test_cap_1, test_cap_2])
+
+
+def service_manifests():
+    """ The agent fixture calls this function when it creates the service manifests
+        that should be used with this test module. The agent fixture expects a list
+        of StandardManifest and/or DefaultManifest objects.
+    """
+    return [manifest]
+
+
+@pytest.fixture
+def create_testoutput_dir(scope="module"):
+    """ Create a directory for the generated test files.
+
+        This directory is ignored by git but it's nice to have
+        somewhere locally to store test output to support
+        troubleshooting etc.
+    """
+    if not os.path.exists(TESTOUTPUT_DIR):
+        os.makedirs(TESTOUTPUT_DIR)
+
+"""##### Test suites #####"""
+
+
+@pytest.mark.usefixtures("testhelper", "create_testoutput_dir", "agent")
+class TestCGroupGateway(object):
+    """ This suite tests that whether CGroupsGateway can working with supported options or not.
+
+        Prerequisites :
+        CONFIG_MEMCG and CONFIG_MEMCG_SWAP kernel options should have been enabled for these tests.
+        Also cgroup_enable=memory and swapaccount=1 options should be enabled by grub.
+    """
+    def test_small_threshold(self):
+        """ Test if the memory limiting with cgroup gateway is working as expected
+            which behavior is when allocated more memory than limit it should fail
+        """
+        try:
+            sc = Container()
+            sc.start(DATA)
+
+            result = sc.set_capabilities(["test.cap.small.threshold"])
+            assert result is False
+        finally:
+            sc.terminate()
+
+    def test_memory_limit(self):
+        """ Test if the memory limiting with cgroup gateway is working as expected
+            which behavior is when allocated more memory than limit it should fail
+        """
+        try:
+            sc = Container()
+            sc.start(DATA)
+
+            result = sc.set_capabilities(["test.cap.memory.share"])
+            assert result is True
+
+            memory_limitation = 1024 * 1024 * 1024
+            _, result = sc.launch_command("python " +
+                                           sc.get_bind_dir() +
+                                           "/testhelper.py" +
+                                           " --test-dir " +
+                                           sc.get_bind_dir() +
+                                           " --do-allocate " +
+                                           str(memory_limitation))
+            assert result is True
+            # wait 5 seconds for previous operation to end
+            time.sleep(5)
+            helper = CGroupHelper(CURRENT_DIR)
+            allocation_return = helper.result()
+            assert allocation_return < memory_limitation
```

Comment:
It can be variable due to kernel space and user space programs allocation it could fail just after 0.9 *maximum  once, and could fail after allocation 0.6 of maximum in another time.